core/src/main/scala/nl/vroste/zio/kinesis/client/zionative/Consumer.scala

```diff
@@ -426,6 +426,95 @@ object Consumer {
            }.runDrain
     } yield ()
 
+  /**
+   * Apply an effectful function to each record in a stream.
+   *
+   * This function ensures that a checkpoint is performed between every 2 records with the same partition key. This is
+   * useful when you want to ensure that records with the same partition key are processed in order. Even in the
+   * presence of failure, only a at most 1 record per partition key will be reprocessed.
+   *
+   * @param streamIdentifier
+   *   Stream to consume from. Either just the name or the whole arn.
+   * @param applicationName
+   *   Name of the application. This is used as the table name for lease coordination (DynamoDB)
+   * @param deserializer
+   *   Record deserializer
+   * @param workerIdentifier
+   *   Identifier of this worker, used for lease coordination. Must be unique for each worker
+   * @param fetchMode
+   *   How to fetch records: Polling or EnhancedFanOut, including config parameters
+   * @param leaseCoordinationSettings
+   *   Config parameters for lease coordination
+   * @param initialPosition
+   *   When no checkpoint exists yet for a shard, start processing from this position
+   * @param emitDiagnostic
+   *   Function that is called for events happening in the Consumer. For diagnostics / metrics
+   * @param shardAssignmentStrategy
+   *   How to assign shards to this worker
+   * @param checkpointBatchSize
+   *   Maximum number of records before checkpointing
+   * @param checkpointDuration
+   *   Maximum interval before checkpointing
+   * @param checkpointRetrySchedule
+   *   Schedule for retrying checkpointing in case of failure
+   * @param processorParallelism
+   *   Number of fibers to process records in parallel per shard
+   * @param recordProcessor
+   *   A function for processing a `Record[T]`
+   * @return
+   *   A ZIO that completes with Unit when record processing is stopped or fails when the consumer stream fails
+   */
+  def consumePartitionedWith[R, RC, T](
```

Comment:
I wonder if it makes more sense to support all of this behavior via an extra parameter in `consumeWith`.

Reply:
May proposal for doing that would be to introduce a new sum type ConsumptionBehaviour and put all the behavior specific paramters in there.

Problem is that this would change signature of consumeWith in a backwards incompatible way. If you are fine with this I can take care of that now, otherwise we can do it for the next major version.

Reply:
Such a change would be source compatible, right? I'm fine with releasing that in a `0.34` version. So far I've been using version numbers below 1.0 as to not make any compatibility guarantees besides best effort, and a bump in the minor version meaning that there are some incompatibilities.

Reply:
The processorParallelism could be useful for the regular consumption behavior as well, if you don't care about ordering and replay behavior, which I can imagine for certain use cases. With two new parameters we would end up with 14 parameters, which we should probably refactor a bit at some point.

Reply:
I'll adjust the setting 👍